lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -568,176 +569,188 @@ def write(chunk)
           tag, group_level_resource)}"
 
         # Does the actual write to the cloud logging api.
-        client = api_client
         if @use_grpc
-          begin
-            labels_utf8_pairs = group_level_common_labels.map do |k, v|
-              [k.encode('utf-8'), convert_to_utf8(v)]
-            end
+          labels_utf8_pairs = group_level_common_labels.map do |k, v|
+            [k.encode('utf-8'), convert_to_utf8(v)]
+          end
 
-            entries_count = entries.length
-            client.write_log_entries(
-              entries,
+          requests_to_send << Google::Logging::V2::WriteLogEntriesRequest.new(
+            entries: entries,
+            log_name: log_name,
+            resource: Google::Api::MonitoredResource.new(
+              type: group_level_resource.type,
+              labels: group_level_resource.labels.to_h
+            ),
+            labels: labels_utf8_pairs.to_h)
+        else
+          requests_to_send << \
+            Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
               log_name: log_name,
-              resource: Google::Api::MonitoredResource.new(
-                type: group_level_resource.type,
-                labels: group_level_resource.labels.to_h
-              ),
-              labels: labels_utf8_pairs.to_h
-            )
-            increment_successful_requests_count
-            increment_ingested_entries_count(entries_count)
-
-            # Let the user explicitly know when the first call succeeded, to aid
-            # with verification and troubleshooting.
-            unless @successful_call
-              @successful_call = true
-              @log.info 'Successfully sent gRPC to Stackdriver Logging API.'
-            end
+              resource: group_level_resource,
+              labels: group_level_common_labels,
+              partial_success: @partial_success,
+              entries: entries)
+        end
+      end
+      requests_to_send.each do |request|
+        client = api_client
+        if @use_grpc
+          write_request_via_grpc(client, request)
```

Comment:
There's got to be a way to just assign a callable value conditionally somewhere at the top and then invoke it here unconditionally.

Reply:
Changed it to use a caller function like this:

```
def write_request_function
  @use_grpc ? 'write_request_via_grpc' : 'write_request_via_http'
end
```

Then
```
send(write_request_function, client, request)
```

WDYT?

Reply:
Hmm, invoke by name (`send`) wouldn't be my first choice. Some alternatives (don't require a separate method):
1. Bound method:
   ```ruby
   write_request = method(if @use_grpc then
                            :write_request_via_grpc
                          else
                            :write_request_via_http
                          end)
   ...
   write_request.call(client, request)
   ```
2. Proc:
   ```ruby
   write_request = if @use_grpc then
                     proc {|v| write_request_via_grpc(v)}
                   else
                     proc {|v| write_request_via_http(v)}
                   end
   ...
   write_request.call(client, request)
   ```
3. Lambda:
   ```ruby
   write_request = if @use_grpc then
                     ->v {write_request_via_grpc(v)}
                   else
                     ->v {write_request_via_http(v)}
                   end
   ...
   write_request.call(client, request)
   ```

Reply:
Changed it to use Bound method.